src/Parser.cpp

```diff
@@ -718,77 +730,66 @@ Parser::Token Parser::next_token() {
   }
 }
 
-void Parser::advance() {
-  PSCM_ASSERT(pos_ <= code_.size());
-  if (code_[pos_] == '\n') {
+void Parser::advance(UChar32 curchar) {
+  if (curchar == '\n') {
     row_++;
     col_ = 0;
   }
-  pos_++;
+  last_token_ += curchar;
   col_++;
 }
 
-bool Parser::is_eof() const {
-  if (use_stream_) {
-    PSCM_ASSERT(in_);
-    return in_->eof();
-  }
-  else {
-    return pos_ >= code_.size();
-  }
-}
-
-char Parser::next_char() {
-  if (use_stream_) {
-    char ch;
-    in_->get(ch);
-    return ch;
-  }
-  else {
-    char ch = code_[pos_];
-    advance();
-    return ch;
-  }
+// 观览器 #4 的辅助类型
+template<class ... Ts>
+struct overloaded : Ts... { using Ts::operator()...; };
+
+/**
+ * 读取当前字符并前移迭代器，类似于 nextposinc 系列。返回EOF表示读到末尾。
+*/
+UChar32 Parser::next_char() {
+  UChar32 ch = std::visit(overloaded{
```

Comment:
src/Parser.cpp:750:27: error: no viable constructor or deduction guide for deduction of template arguments of 'overloaded'